### DISCUSSION TITLE
Support for BitBucket deployment environments

### DISCUSSION THREAD

**Discussion by Radiergummi** [Infisical/infisical#6462] [Feature Request]
### Feature description
Right now, the BitBucket integration allows to connect an Infisical environment to a single repository, which will sync all secrets into the repository configuration.  
However, this doesn't really mirror the way software teams usually work: For pretty much all projects I've seen before, there are separate branches for development, staging and production.

BitBucket has a hierarchical variable system, with finer-grained variables overriding those in a broader scope:
1. Workspace variables (workspaces are roughly similar to GitHub orgs)
2. Repository variables
3. Deployment environment variables

This means that if a variable `FOO` is specified in all of a workspace, a repository, and the deployment environment _staging_, the definition in the _staging_ deployment environment "wins".  

I'd like to suggest adding support for BitBucket deployment environments to the integration, such that I can actually specify staging and production secrets in Infisical and use them in the respective deployment pipeline.  
This would require adding an additional dropdown to the integration creation screen, probably named "BitBucket Deployment environment", which could be either one of the existing environments on BitBucket, or none to sync with the repository variables as it is currently implemented.

### Why would it be useful?
A deployment environment can be used in Bitbucket Pipelines to dedicate deployment steps in a pipeline configuration to a named deployment target. For example, in the build configuration of my "staging" branch, I can define a step such as:
```yaml
branches:
  staging:
    steps:
      - name: Deploy
        deployment: staging  # <-- deployment environment here
        script:
          - gcloud container clusters get-credentials "${CLUSTER_NAME}"
                                                       # ^-------------- variable
```
and be sure `CLUSTER_NAME` will resolve to the variable I defined in the "staging" environment as referenced in the step.

The UI for them looks like so:
<img width="638" alt="Bildschirmfoto 2023-07-31 um 12 14 37" src="https://github.com/Infisical/infisical/assets/6115429/96f917b9-6abb-4dd7-8471-7bc6c76ea60b">

### Additional information
I haven't looked at your source code yet, but API-wise, the deployment environments are just another layer of resources with their own endpoints:
- [List deployment environments](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-deployments/#api-repositories-workspace-repo-slug-environments-get)
- [List variables for a deployment environment](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pipelines/#api-repositories-workspace-repo-slug-deployments-config-environments-environment-uuid-variables-get)

With the environment UUID, you can just prefix all API calls to create/update/delete secrets with the environment, and it should work fine.

**Comment by Salman2301** [comment]
I am happy to work on this looks same as GitHub integrations #1398 PR
